engine/src/main/java/org/hibernate/search/engine/impl/ImmutableSearchFactory.java

```diff
@@ -131,15 +131,20 @@ public ImmutableSearchFactory(SearchFactoryState state) {
 		this.timeoutExceptionFactory = state.getDefaultTimeoutExceptionFactory();
 		this.timingSource = state.getTimingSource();
 		this.mapping = state.getProgrammaticMapping();
-		this.statistics = new StatisticsImpl( this );
+		if ( state.getStatistics() == null ) {
+			this.statistics = new StatisticsImpl( this );
+			boolean statsEnabled = ConfigurationParseHelper.getBooleanValue(
+					configurationProperties, Environment.GENERATE_STATS, false
+			);
+			this.statistics.setStatisticsEnabled( statsEnabled );
+		}
+		else {
+			this.statistics = (StatisticsImpl) state.getStatistics();
```

Comment:
The cast is safe because `StatisticsImpl` is created only in the `ImmutableSearchFactory`'s constructor itself, 4 rows before the cast.

Otherwise, setting up `StatisticsImpl` as the return type of the the `SearchFactoryState.getStatistics()` method avoids the use of the cast but, honestly, i prefer to have the `Statistics` interface as return type rather than a class implementing that interface (i.e. `StatisticsImpl`).
Anyway, if we change `SearchFactoryState.getStatistics()`'s return type to `StatisticsImpl`, we have to change the `MutableSearchFactoryState` class that implements `SearchFactoryState` and the two factories `ImmutableSearchFactory` and `MutableSearchFactory`.
In this way in the `ImmutableSearchFactory` class constructor there's no need to cast anymore.